presto-main/src/main/java/com/facebook/presto/application/ApplicationManager.java

```diff
@@ -0,0 +1,114 @@
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class ApplicationManager
+{
+    private static final Logger log = Logger.get(ApplicationManager.class);
+
+    @GuardedBy("this")
+    private final Map<String, ApplicationFactory> applicationFactories = new ConcurrentHashMap<>();
+    @GuardedBy("this")
+    private final Map<String, Application> applications = new ConcurrentHashMap<>();
+
+    private final AtomicBoolean stopped = new AtomicBoolean();
+
+    public synchronized void addApplicationFactory(ApplicationFactory applicationFactory)
+    {
+        requireNonNull(applicationFactory, "applicationFactory is null");
+
+        if (applicationFactories.putIfAbsent(applicationFactory.getName(), applicationFactory) != null) {
+            throw new IllegalArgumentException(format("Application '%s' is already registered", applicationFactory.getName()));
+        }
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        if (stopped.getAndSet(true)) {
+            return;
+        }
+
+        for (Map.Entry<String, Application> entry : applications.entrySet()) {
+            Application application = entry.getValue();
+            try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(application.getClass().getClassLoader())) {
+                application.shutdown();
+            }
+            catch (Throwable t) {
+                log.error(t, "Error shutting down application: %s", entry.getKey());
+            }
+        }
+    }
+
+    protected synchronized void setup(String applicationName, ApplicationFactory applicationFactory, Map<String, String> properties)
+    {
+        checkState(!stopped.get(), "ApplicationManager is stopped");
+        checkState(!applications.containsKey(applicationName), "Application %s already exists", applicationName);
+        try {
+            applications.put(applicationName, applicationFactory.create(properties));
+        }
+        catch (Throwable t) {
+            log.error(t, "Error initializing application: %s", applicationFactory.getName());
+        }
+    }
+
+    public void setup(String applicationName, Map<String, String> properties)
+    {
+        checkState(!stopped.get(), "ApplicationManager is stopped");
+        ApplicationFactory applicationFactory = applicationFactories.get(applicationName);
+        checkArgument(applicationFactory != null, "No factory for application %s", applicationName);
+        setup(applicationName, applicationFactory, properties);
+    }
+
+    protected void start(Injector injector)
+    {
+        for (Map.Entry<String, Application> entry : applications.entrySet()) {
+            Application application = entry.getValue();
+            try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(application.getClass().getClassLoader())) {
+                application.run(injector);
```

Comment:
Passing `Injector` here won't work, since plugins can't see Presto's Guice from inside their class loader, by design. We intentionally do not expose the internal details of Presto to plugins. Anything needed by a plugin service should be explicitly part of the SPI.